Add a default provider and implement SHA256 in it

mattcaswell opened the pull request:
This is marked as WIP because it still needs tests and documentation. It is also built on the commits in #8341 and #8480.

This runs through "make test" successfully using the default provider implementation of SHA256 where applicable. If engines are involved or we are invoking the EVP_Digest* functions as part of an EVP_DigestSign* operation then we fallback to the legacy handling.

levitte commented on crypto/provider_predefined.h:
```diff
@@ -9,13 +9,15 @@
 
+int ossl_default_provider_init(const OSSL_PROVIDER *provider,
+                               const OSSL_DISPATCH *in,
+                               const OSSL_DISPATCH **out);
```
You can actually do this:

``` C
OSSL_provider_init_fn ossl_default_provider_init;
```

mattcaswell replied:
I rebased on top of your latest version of #8480 - so picked this up from your branch.

mattcaswell commented:
Rebased now that #8341 went in.

paulidale commented on providers/common/digests/sha.c:
```diff
@@ -0,0 +1,76 @@
+#include <openssl/core_numbers.h>
+
+
+static int sha256_init(void *ctx)
```
Init and update could go directly couldn't they?
Rather than having a function wrap them.

mattcaswell replied:
Good point! Done.

paulidale commented on providers/common/digests/sha.c:
```diff
@@ -0,0 +1,76 @@
+{
+    SHA256_CTX *ctx = (SHA256_CTX *)vctx;
+
+    if (ctx != NULL)
```
`OPENSSL_clear_free` instead of these three lines?

mattcaswell replied:
Done

paulidale commented on providers/default/defltprov.c:
```diff
@@ -0,0 +1,114 @@
+    return deflt_param_types;
+}
+
+static int deflt_get_params(const OSSL_PROVIDER *prov, OSSL_PARAM params[])
```
`params` should be const should't it?

mattcaswell replied:
Yes. Fixed.

paulidale commented on providers/default/defltprov.c:
```diff
@@ -0,0 +1,114 @@
+    static const char provider_name[] = "OpenSSL Default Provider";
+
+    for (; ok && p->key != NULL; p++) {
+        if (strcmp(p->key, OSSL_PROV_PARAM_NAME) == 0) {
```
The _params.h_ helpers?

```
return OSSL_PARAM_set_utf8_string(OSSL_PARAM_locate(params, OSSL_PROV_PARAM_NAME), provider_name))
    && OSSL_PARAM_set_utf8_string(OSSL_PARAM_locate(params, OSSL_PROV_PARAM_VERSION), OPENSSL_VERSION_STR))
    && OSSL_PARAM_set_utf8_string(OSSL_PARAM_locate(params, OSSL_PROV_PARAM_BUILDINFO), OPENSSL_FULL_VERSION_STR));
```

These do buffer size checking, return size setting etc.

paulidale replied:
I think the combined search and get/set helpers have a place still.
This kind of sequence is going to appear a lot thought out the code.

paulidale replied:
New locating helpers in #8519

levitte replied:
Your use of the params.h functions changes the semantics. What happens if the other end only asks for `OSSL_PROV_PARAM_VERSION`?

To preserve semantics, it should be something like this, no? 

``` C
    const OSSL_PARAM *p;

    return ((p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_NAME)) == NULL || OSSL_PARAM_set_utf8_string(p, provider_name)) 
        && ((p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_VERSION)) == NULL || OSSL_PARAM_set_utf8_string(p, OPENSSL_VERSION_STR)) 
        && ((p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_BUILDINFO)) == NULL || OSSL_PARAM_set_utf8_string(p, OPENSSL_FULL_VERSION_STR)) 
```

paulidale replied:
My bad.  I forgot a NULL param returned error.  Maybe it should return success in this case.

With the new functions, this would be:

```
return OSSL_PARAM_locate_set_utf8_string(params, OSSL_PROV_PARAM_NAME, provider_name))
    && OSSL_PARAM_locate_set_utf8_string(params, OSSL_PROV_PARAM_VERSION, OPENSSL_VERSION_STR))
    && OSSL_PARAM_locate_set_utf8_string(params, OSSL_PROV_PARAM_BUILDINFO, OPENSSL_FULL_VERSION_STR));
```

That the original missed the length checks, type checks and the return size is a concern about the complexity involved :(

mattcaswell replied:
Updated based on @paulidale's latest suggestion

paulidale commented on providers/default/defltprov.c:
```diff
@@ -0,0 +1,114 @@
+
+/* Parameters we provide to the core */
+static const OSSL_ITEM deflt_param_types[] = {
+    { OSSL_PARAM_UTF8_STRING, OSSL_PROV_PARAM_NAME },
```
These could be `OSSL_PARAM_UTF8_PTR`, avoiding a copy.

mattcaswell replied:
Done

paulidale commented on crypto/evp/digest.c:
```diff
@@ -296,3 +483,99 @@ int EVP_MD_CTX_ctrl(EVP_MD_CTX *ctx, int cmd, int p1, void *p2)
+    if ((md = EVP_MD_meth_new(mdtype, NID_undef)) == NULL)
+        return NULL;
+
+    for (; fns->function_id != 0; fns++) {
```
I'm not sure how, but this is begging to be made table driven.

mattcaswell replied:
Left as an exercise for the future. :-)

levitte commented on doc/man3/EVP_MD_fetch.pod:
```diff
@@ -0,0 +1,146 @@
+provider. The names of these algorithms are: RIPEMD160, MD2, MD4, MD5, MDC2 and
+whirlpool.
+
+The B<properties> parameter specifies the search criteria that will be used to
```
It would be possible with Pod::Templates... Alas, that's not part of the perl core.

mattcaswell replied:
My original idea was that all *fetch* functions would end up on this same man page and we would adjust the wording to be generic for all of them. But even if we don't end up doing that I'd suggest deferring the issue until we need it in more than one place.

mattcaswell commented on crypto/evp/digest.c:
```diff
@@ -163,14 +282,40 @@ int EVP_DigestFinal(EVP_MD_CTX *ctx, unsigned char *md, unsigned int *size)
     int ret;
+    size_t size = 0;
+
+    if (ctx->digest->prov == NULL)
```
Fixed.

mattcaswell commented on crypto/evp/digest.c:
```diff
@@ -163,14 +282,40 @@ int EVP_DigestFinal(EVP_MD_CTX *ctx, unsigned char *md, unsigned int *size)
+    ret = ctx->digest->dfinal(ctx->provctx, md, &size);
+
+    if (isize != NULL) {
+        if (size <= INT_MAX) {
```
Fixed.

mattcaswell commented on crypto/evp/digest.c:
```diff
@@ -296,3 +483,99 @@ int EVP_MD_CTX_ctrl(EVP_MD_CTX *ctx, int cmd, int p1, void *p2)
+    for (; fns->function_id != 0; fns++) {
+        switch (fns->function_id) {
+        case OSSL_FUNC_DIGEST_NEWCTX:
+            if (md->newctx != NULL)
```
Not sure. I chose to tolerate this, but maybe that's wrong.

mattcaswell commented on doc/man3/EVP_MD_fetch.pod:
```diff
@@ -0,0 +1,146 @@
+whirlpool.
+
+The B<properties> parameter specifies the search criteria that will be used to
+look for an algorithm implementation. Properties are given as a space delimited
```
Fixed.

mattcaswell commented on doc/man3/EVP_MD_fetch.pod:
```diff
@@ -0,0 +1,146 @@
+have the property "fips=yes". In the event that more than one implementation
+of the given algorithm name matches the specified properties then any one of
+those implementations may be returned. The B<properties> parameter may be NULL
+in which case any implementation from the available providers with the given
```
Done

mattcaswell commented on doc/man3/EVP_MD_fetch.pod:
```diff
@@ -0,0 +1,146 @@
+ /* This only needs to be done once - usually at application start up */
+ OSSL_PROVIDER *legacy = OSSL_PROVIDER_load(NULL, "legacy");
+
+ EVP_MD *md = EVP_MD_fetch(NULL, "whirlpool", "legacy=yes");
```
I added that. I also added an example showing having both the legacy and default providers loaded at the same time.

mattcaswell commented:
I rebased this now that #8480 has gone in. This incorporates all fixes to the feedback so far. The previously noted dependencies for this PR (#8341 and #8480) have both been merged, but now we additionally depend on #8519.

levitte commented on providers/default/defltprov.c:
```diff
@@ -0,0 +1,98 @@
+                                             OPENSSL_VERSION_STR)
+           && OSSL_PARAM_locate_set_utf8_ptr(params,
+                                             OSSL_PROV_PARAM_BUILDINFO,
+                                             OPENSSL_FULL_VERSION_STR);
```
Using `OSSL_PARAM_locate_set_utf8_ptr` forces everyone that's interested in any of these parameters to ask for all of them.  Please consider using `OSSL_PARAM_optional_set_utf8_ptr` instead.

(a final argument should be that if the provider is updated to contain a new parameter in this set, every application using the new version will fail because they haven't updated their end to ask for the new parameter as well.  *That*.  *Would*.  *Be*.  *Really*.  *Bad*.)

mattcaswell replied:
Done

richsalz commented on providers/default/defltprov.c:
```diff
@@ -0,0 +1,98 @@
+    return OSSL_PARAM_optional_set_utf8_ptr(params,
+                                            OSSL_PROV_PARAM_NAME,
+                                            "OpenSSL Default Provider")
+           && OSSL_PARAM_optional_set_utf8_ptr(params,
```
This stops if one fails.  Don't you want to try to do them all?

mattcaswell replied:
I'd say no. These function succeed if a given parameter is absent from from params, so a failure would have to be something like the datatype doesn't match what we're expecting. That's probably a legitimate failure because it shouldn't happen.

levitte replied:
Agreed.  If one of the OSSL_PARAM items has the wrong data_type, the get_params operation should fail.  Sure, it means that *some* of them still have values, but that's inconsequential.

paulidale replied:
Resolving with the closing of #8519

mattcaswell commented:
I've rebased this and removed the dependency on #8519 (which got closed without merge). So this PR is now completely standalone.

I've taken it out of WIP. Please take a look.

paulidale commented on crypto/evp/evp_lib.c:
```diff
@@ -316,6 +317,10 @@ int EVP_MD_size(const EVP_MD *md)
+
+    if (md->prov != NULL && md->size != NULL)
+        return (int)md->size();
+
```
Does this need a legacy to be removed comment?
The _if_ above looks like it will change too.

mattcaswell replied:
No - this is non-legacy code!

paulidale commented on doc/man3/EVP_MD_fetch.pod:
```diff
@@ -0,0 +1,162 @@
+With implicit lookup an application can use functions such as L<EVP_sha256(3)>,
+L<EVP_sha512(3)> or L<EVP_blake2b512(3)> to obtain an B<EVP_MD> object. When
+used in a function like L<EVP_DigestInit_ex(3)> the actual implementation to
+be used will be fetched implicitly using default search criteria. Typically
```
Comma after _typically_?

mattcaswell replied:
Done

paulidale commented on doc/man3/EVP_MD_fetch.pod:
```diff
@@ -0,0 +1,162 @@
+ EVP_MD_meth_free(md);
+
+Note that in the above example the property string "legacy=yes" is optional
+since (assuming no other providers have been loaded) the only implmentation of
```
Commas instead of parenthesis?

mattcaswell replied:
Done

paulidale commented on providers/common/digests/sha.c:
```diff
@@ -0,0 +1,63 @@
+/*
```
Is it worthwhile naming this file `sha2.c`?
Apart from the names, there isn't a lot of similarity between SHA1, SHA2 and SHA3.

mattcaswell replied:
Done

paulidale commented on providers/default/defltprov.c:
```diff
@@ -0,0 +1,102 @@
+    { 0, NULL }
+};
+
+int ossl_default_provider_init(const OSSL_PROVIDER *provider,
```
This prototype should be removed.

levitte replied:
Ideally yes.  Unfortunately, `--strict-warnings` forces us to have prototypes for all non-static functions.

levitte replied:
The prototype could be written like this, though:

``` C
OSSL_provider_init_fn ossl_default_provider_init;
```

That *would* work to ensure that the function gets defined correctly.

mattcaswell replied:
Right - as @levitte says unfortunately we need the prototype there. I have rewritten it in the form suggested though.

paulidale replied:
Sounds good. It won't look so glaring written the other way.

If it isn't static, it should be in a header somewhere shouldn't it?

mattcaswell replied:
Could be. But since there is exactly one consumer I didn't feel a great need to.

paulidale commented on include/openssl/core_numbers.h:
```diff
@@ -72,6 +72,32 @@ OSSL_CORE_MAKE_FUNC(const OSSL_ALGORITHM *,provider_query_operation,
+# define OSSL_FUNC_DIGEST_DIGEST            5
+# define OSSL_FUNC_DIGEST_FREECTX           6
+# define OSSL_FUNC_DIGEST_DUPCTX            7
+# define OSSL_FUNC_DIGEST_SIZE              8
```
Will a ctrl function be needed in due course?  e.g. SHAKE's output size.

levitte replied:
That could be the job of a algorithm level set_params function.

mattcaswell replied:
We will need something like that, but I've deliberately deferred dealing with that until we need to.

mattcaswell commented:
Fixup commits pushed. @paulidale please can you reconfirm?

paulidale commented:
AppVeyor failure isn't relevant.

levitte commented:
I restarted the failed appveyor build

mattcaswell commented:
Appveyor passed this time.

mattcaswell commented:
Pushed. Thanks.
